### PR TITLE
Add a function to build images immediately

### PR DESCRIPTION
**Pull request by pawalt**
## Describe your changes

Resolves MOD-4102

<details> <summary>Backward/forward compatibility checks</summary>

---

Check these boxes or delete any item (or this section) if not relevant for this PR.

- [x] Client+Server: this change is compatible with old servers
- [x] Client forward compatibility: this change ensures client can accept data intended for later versions of itself

Note on protobuf: protobuf message changes in one place may have impact to
multiple entities (client, server, worker, database). See points above.

---

</details>

## Changelog

Images can now be built directly with `Image.build`.

### CODE REVIEW

File: modal/image.py [hunk 1]
```diff
@@ -465,6 +467,15 @@ async def join():
         obj.force_build = force_build
         return obj
 
+    async def build(self, app: "_App", client: Optional[_Client] = None) -> None:
+        """Build the image immediately rather than waiting for it to be used."""
+
+        client = client or app._client or await _Client.from_env()
```
Comment (mwaskom) [inline modal/image.py:473]: Do we need `app` as an argument here? Feels confusing to have an object that is usually passed to `App` have a method that accepts an `App` as an argument.

File: modal/image.py [hunk 2]
```diff
@@ -465,6 +467,15 @@ async def join():
         obj.force_build = force_build
         return obj
 
+    async def build(self, app: "_App", client: Optional[_Client] = None) -> None:
+        """Build the image immediately rather than waiting for it to be used."""
```
Comment (mwaskom) [inline modal/image.py:471]: Nit but the "waiting for it to be used" here is arguably not a correct description of what usually happens — images get build when we create apps or functions that _will_ use them.

**Comment by erikbern** [comment]
I think `image.resolve()` might already do this, but could be wrong! Either way this I think this points to some complexity around managing "lazy" vs "eager" objects (c.f. `from_name` vs `lookup`)

**Review by mwaskom — CHANGES_REQUESTED** [review]
Discussed with Peyton on Slack but I don't think this actually solves the user's problem — not sure we should expand API without a clear rationale.

**Comment by pawalt** [comment]
Cancelling as we have better ways to solve this UX pain

**Pull request closed without merging**